### DISCUSSION TITLE
[Self-Hosting] esm.sh is incompatible with Azure DevOps (private) npm registries

### DISCUSSION THREAD

**Discussion by MilanKovacic** [esm-dev/esm.sh#701] [General]
Hello!
Azure DevOps API does not support fetching /package/version. It only supports /package endpoint, so all of the subsequent requests fail.
Since /package response is a superset of /package/version, information could be extracted from the response.
Currently, the only workaround is to use an anti-corruption layer to proxy the requests and manipulate the response for /package/version endpoint.

Are there any plans to officially support Azure DevOps? Thank you in advance.

**Comment by ije** [comment]
didn't get the `/package/version`? can you please give a specific example?

**Reply by MilanKovacic** [reply]
Sure. The official npm registry provides these two endpoints. Azure DevOps npm registry only supports the first endpoint. For example, calling [AzureDevOpsRegistry]/chalk works, while [AzureDevOpsRegistry]/chalk/5.3.0 does not. To replicate the npm experience with Azure API for this example, one would have to first call [AzureDevOpsRegistry]/chalk to obtain the full Json (containing all versions), and then extract the specific version.

#### `GET·/{package}`

| Name     | Value     | Kind     | Required?     | Notes     |
|------    |-------    |------    |-----------    |-------    |
| package | String | **Path**  | ✅         | the name of the package |

This endpoint responds with the package metadata document, sometimes informally called a "packument" or "doc.json". The format of the response is described in detail in the [package metadata documentation](responses/package-metadata.md).

#### `GET·/{package}/{version}`

| Name     | Value     | Kind     | Required?     | Notes     |
|------    |-------    |------    |-----------    |-------    |
| package | String | **Path** | ✅         | the name of the package |
| version | String | **Path** | ✅         | a version number or `latest`        |

**Reply by MilanKovacic** [reply]
One approach JSPM takes is a plugin-based system:

https://jspm.org/docs/0.16/registry-api.html
